Include/moduleobject.h

```diff
@@ -90,6 +94,13 @@ struct PyModuleDef_Slot {
 #  define Py_MOD_PER_INTERPRETER_GIL_SUPPORTED ((void *)2)
 #endif
 
+/* for Py_mod_gil: */
+#if !defined(Py_LIMITED_API) || Py_LIMITED_API+0 >= 0x030e0000
+#  define Py_MOD_GIL_USED ((void *)0)
+#  define Py_MOD_GIL_NOT_USED ((void *)1)
+PyAPI_FUNC(int) PyModule_SetGIL(PyObject *module, void *gil);
+#endif
```

Comment:
Presumably this is so single-phase init modules can indicate they do or do not support running without a GIL.  To me, it make more sense to say that all single-phase init modules require the GIL.

Reply:
> Presumably this is so single-phase init modules can indicate they do or do not support running without a GIL

Yeah, it's for gh-117526. I put it in this PR because it ended up being such a small change.

> it make more sense to say that all single-phase init modules require the GIL

I don't think we should make that assumption. @colesbury said that when he did a survey of popular numerical computing packages (like NumPy and PyTorch), most of them still used single-phase init. Requiring developers to move to multi-phase init to support free-threading could be an unreasonable burden; there's no reason why global state can't be made thread-safe while remaining global state (and not supporting subinterpreters).

> is this really supposed to go in the limited API?

I was modeling it after the existing subinterpreters slot, but it doesn't have to be in the limited API. I can take it out of the limited API for now and we can discuss adding it back later if a need comes up.

Are you talking about `PyModule_SetGIL()` specifically or everything related to the `Py_mod_gil` slot?

Reply:
> > it make more sense to say that all single-phase init modules require the GIL
> 
> I don't think we should make that assumption

Changing a module to multi-phase init takes small effort *if* you isolate the module between interpreters (e.g. don't port static types or switch to module state).  Along with that minimal effort, it should then be sufficient to set the `Py_mod_multiple_interpreters` slot to `Py_MOD_MULTIPLE_INTERPRETERS_NOT_SUPPORTED`.  That's a stopgap that would allow people to use the `Py_mod_gil` slot to indicate they support free-threading (or don't).  Then we wouldn't need to add `PyModule_SetGIL()` to accomplish the same thing.

(That said, encouraging folks to implement multi-phase init without implementing interpreter isolation isn't ideal.  I'm sure @encukou has an opinion. 😄)

> Are you talking about `PyModule_SetGIL()` specifically or everything related to the `Py_mod_gil` slot?

I was talking about `PyModule_SetGIL()` specifically.  Putting the new module slot values in the limited API is fine.

Reply:
Opinion? I have a [documented solution](https://docs.python.org/3/howto/isolating-extensions.html#opt-out-limiting-to-one-module-object-per-process) that predates `Py_mod_multiple_interpreters` ;)
Alas, it relies on the GIL...

Reply:
`PyModule_SetGIL()` does not need to be part of the limited API. Unfortunately, the limited API doesn't currently work with free-threading, so there isn't much reason to include it as part of the limited API.

I feel pretty strongly that we should not require package maintainers to rewrite their extensions to use multi-phase init, just to support free-threading. Even without isolating state, it's a bunch of extra work that is not necessary. Many packages have multiple C API modules (NumPy has [~15](https://github.com/search?q=repo%3Anumpy%2Fnumpy%20PyMODINIT_FUNC&type=code)) so it would require a substantial amount of refactoring.

Reply:
> I feel pretty strongly that we should not require package maintainers to rewrite their extensions to use multi-phase init,

The level of effort would be minimal:

1. move the content of each init function to a corresponding "module exec" function (drop call to create module object)
2. add a static `PyModuleDef_Slot` var for each, with a `Py_mod_exec` slot
3. add the `Py_mod_multiple_interpreters`, set to `Py_MOD_MULTIPLE_INTERPRETERS_NOT_SUPPORTED`
4. set `def.m_size` to 0 (if negative)
5. update each init function to only `return PyModuleDef_Init(def);` for the corresponding module def

There's no need to do any of the other stuff (heap types, module state, etc.) immediately.

At that point, the only extra step to run on a free-threading build would be to add the `Py_mod_gil` slot, set to `Py_MOD_GIL_USED `.

As to single-phase init modules, we'd always treat them as though they had `Py_MOD_GIL_USED `.

> just to support free-threading.

We want people to switch to multi-phase init regardless.  Single-phase init modules continue to be a pain point for the CPython runtime, as well as limiting options (e.g. reload, subinterpreters) for the extension modules and requiring that they do work that the import machinery could do for them.

Free-threading support would be yet another carrot.

> Even without isolating state, it's a bunch of extra work that is not necessary.

The steps I noted above should be all that's necessary and I would not expect them to take more than a matter of minutes for each module (whether one in a file or many).  Perhaps I've missed something?

> it would require a substantial amount of refactoring.

Likewise, I'm not clear on what substantial refactoring would be necessary.

Reply:
My main concern is that we'd be adding `PyModule_SetGIL()` when we already have "one obvious way to do it": multi-phase init with the `Py_mod_gil` slot.  (Furthermore, I'd rather we not make any new accommodations for people to stay on single-phase init.)

If you still have a strong opinion in favor of `PyModule_SetGIL()` then I recommend opening a thread on DPO to get feedback from a wider audience.  I realize time is short for the 3.13 feature freeze; perhaps we can find a middle-ground in the next few days, or convince @Yhg1s to let us add `PyModule_SetGIL()` after beta 1 (assuming we end up adding it).

In the meantime, could we land this PR without `PyModule_SetGIL()`?

Reply:
@ericsnowcurrently our concern with *requiring* multi-phase init for free-threading support is effectively crippling the PEP 703 effort. Our latest estimates are that close to 90% of existing extensions on PyPI do not in fact use multi-phase init at this point. We won't be able to get much data from them if we put a big yak shave before they can test free threading.

Reply:
If it's just a temporary thing so people can more easily try it out then could we at least call the function something more obvious and less likely to inhibit the move to multi-phase init, like `PyModule_ExperimentalSetGIL()`, and make it available only on `Py_DISABLE_GIL` builds?

Reply:
The `PyModule_ExperimentalSetGIL` name and `Py_DISABLE_GIL` guard sounds fine to me.

Reply:
> call the function something more obvious and less likely to inhibit the move to multi-phase init, like `PyModule_ExperimentalSetGIL()`, and make it available only on `Py_DISABLE_GIL` builds?

Yeah, I can definitely do that. There's also the [existing convention for marking C-API functions as unstable](https://docs.python.org/3/c-api/stable.html#unstable-c-api), so another option is `PyUnstable_Module_SetGIL()`. Which one would you prefer?

I'm currently working on the updates described in [this comment](https://github.com/python/cpython/pull/116882#issuecomment-2086475722), and I'll include your suggested change.

Reply:
`PyUnstable_` sounds good, but I think it would also still help to make it extra obvious: `PyUnstable_Module_ExperimentalSetGIL()`.  If that's too much then I'd rather `PyModule_ExperimentalSetGIL()`.